File: silverwind-frontend/src/app/core/services/asset.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { environment } from '../../../environments/environment';
import { Asset, AssetAssignment, AssetCondition } from '../models/asset.model';
import { ApiResponse } from '../models/api-response.model';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';

export interface PageResponse<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
}

@Injectable({
  providedIn: 'root',
})
export class AssetService {
  private http = inject(HttpClient);
  private apiUrl = `${environment.apiUrl}/assets`;

  // ============ ASSET CRUD (Admin) ============

  listAssets(query?: string, page = 0, size = 10): Observable<PageResponse<Asset>> {
    let params = new HttpParams().set('page', page).set('size', size);
    if (query) params = params.set('q', query);
    return this.http
      .get<ApiResponse<PageResponse<Asset>>>(this.apiUrl, { params })
      .pipe(map((response) => response.data));
  }

  getAsset(id: string): Observable<Asset> {
    return this.http
      .get<ApiResponse<Asset>>(`${this.apiUrl}/${id}`)
      .pipe(map((response) => response.data));
  }

  createAsset(asset: Partial<Asset>): Observable<Asset> {
    let params = new HttpParams()
      .set('assetTag', asset.assetTag!)
      .set('assetType', asset.assetType!);
    if (asset.brand) params = params.set('brand', asset.brand);
    if (asset.model) params = params.set('model', asset.model);
    if (asset.serialNumber) params = params.set('serialNumber', asset.serialNumber);
    if (asset.purchaseDate) params = params.set('purchaseDate', asset.purchaseDate);
    if (asset.warrantyUntil) params = params.set('warrantyUntil', asset.warrantyUntil);
    if (asset.active !== undefined) params = params.set('active', asset.active);
    if (asset.totalQuantity) params = params.set('totalQuantity', asset.totalQuantity);

    return this.http
      .post<ApiResponse<Asset>>(this.apiUrl, null, { params })
      .pipe(map((response) => response.data));
  }

  updateAsset(id: string, asset: Partial<Asset>): Observable<Asset> {
    let params = new HttpParams();
    if (asset.assetType) params = params.set('assetType', asset.assetType);
    if (asset.brand) params = params.set('brand', asset.brand);
    if (asset.model) params = params.set('model', asset.model);
    if (asset.serialNumber) params = params.set('serialNumber', asset.serialNumber);
    if (asset.purchaseDate) params = params.set('purchaseDate', asset.purchaseDate);
    if (asset.warrantyUntil) params = params.set('warrantyUntil', asset.warrantyUntil);
    if (asset.active !== undefined) params = params.set('active', asset.active);
    if (asset.totalQuantity) params = params.set('totalQuantity', asset.totalQuantity);

    return this.http
      .put<ApiResponse<Asset>>(`${this.apiUrl}/${id}`, null, { params })
      .pipe(map((response) => response.data));
  }

  deleteAsset(id: string): Observable<void> {
    return this.http.delete<ApiResponse<void>>(`${this.apiUrl}/${id}`).pipe(map(() => void 0));
  }

  // ============ ASSIGNMENT (Admin) ============

  assignAsset(
    assetId: string,
    userId: string,
    assignedOn?: string,
    condition?: AssetCondition,
    notes?: string,
  ): Observable<AssetAssignment> {
    let params = new HttpParams();
    if (assignedOn) params = params.set('assignedOn', assignedOn);
    if (condition) params = params.set('condition', condition);
    if (notes) params = params.set('notes', notes);

    return this.http
      .post<ApiResponse<AssetAssignment>>(`${this.apiUrl}/${assetId}/assign/${userId}`, null, {
        params,
      })
      .pipe(map((response) => response.data));
  }

  getAssetHistory(assetId: string): Observable<AssetAssignment[]> {
    return this.http
      .get<ApiResponse<AssetAssignment[]>>(`${this.apiUrl}/${assetId}/history`)
      .pipe(map((response) => response.data));
  }

  confirmReturn(
    assignmentId: string,
    returnedOn?: string,
    conditionOnReturn?: AssetCondition,
    notes?: string,
  ): Observable<AssetAssignment> {
    let params = new HttpParams();
    if (returnedOn) params = params.set('returnedOn', returnedOn);
    if (conditionOnReturn) params = params.set('conditionOnReturn', conditionOnReturn);
    if (notes) params = params.set('notes', notes);

    return this.http
      .post<
        ApiResponse<AssetAssignment>
      >(`${this.apiUrl}/assignments/${assignmentId}/confirm-return`, null, { params })
      .pipe(map((response) => response.data));
  }

  // ============ EMPLOYEE SELF-SERVICE ============

  getMyAssets(): Observable<AssetAssignment[]> {
    return this.http
      .get<ApiResponse<AssetAssignment[]>>(`${this.apiUrl}/my-assets`)
      .pipe(map((response) => response.data));
  }

  requestReturn(assignmentId: string): Observable<AssetAssignment> {
    return this.http
      .post<
        ApiResponse<AssetAssignment>
      >(`${this.apiUrl}/assignments/${assignmentId}/request-return`, null)
      .pipe(map((response) => response.data));
  }
}

File: silverwind-backend/src/main/java/com/solventek/silverwind/org/AssetService.java
package com.solventek.silverwind.org;

import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.solventek.silverwind.auth.Employee;
import com.solventek.silverwind.auth.EmployeeRepository;
import com.solventek.silverwind.enums.AssetAssignmentStatus;
import com.solventek.silverwind.enums.AssetCondition;
import com.solventek.silverwind.security.UserPrincipal;

import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Service
@RequiredArgsConstructor
@Transactional
@Slf4j
public class AssetService {

    private final AssetRepository assetRepository;
    private final EmployeeAssetAssignmentRepository assignmentRepository;
    private final EmployeeRepository employeeRepository;
    private final OrganizationRepository organizationRepository;
    private final com.solventek.silverwind.timeline.TimelineService timelineService;
    private final com.solventek.silverwind.notifications.NotificationService notificationService;

    // ============ ASSET CRUD ============

    public Asset createAsset(String assetTag, String assetType, String brand, String model,
            String serialNumber, LocalDate purchaseDate, LocalDate warrantyUntil,
            Boolean active, Integer totalQuantity) {
        log.info("Attempting to create asset: {} for org", assetTag);
        try {
            UUID orgId = getCurrentUserOrgId();

            if (assetRepository.existsByAssetTagAndOrganization_Id(assetTag, orgId)) {
                log.warn("Asset creation failed: Asset tag {} already exists for org {}", assetTag, orgId);
                throw new IllegalArgumentException("Asset tag already exists in this organization");
            }

            Organization org = organizationRepository.findById(orgId)
                    .orElseThrow(() -> new EntityNotFoundException("Organization not found"));

            Asset asset = Asset.builder()
                    .assetTag(assetTag)
                    .assetType(assetType)
                    .brand(brand)
                    .model(model)
                    .serialNumber(serialNumber)
                    .purchaseDate(purchaseDate)
                    .warrantyUntil(warrantyUntil)
                    .active(active != null ? active : true)
                    .totalQuantity(totalQuantity != null ? totalQuantity : 1)
                    .organization(org)
                    .build();

            Asset saved = assetRepository.save(asset);
            log.info("Asset created successfully: {}", saved.getId());

            timelineService.createEvent(orgId, "ASSET", saved.getId(), "CREATE", "Asset Created", getCurrentUserId(),
                    "Asset " + assetTag + " created", null);
            return saved;
        } catch (Exception e) {
            log.error("Error creating asset {}: {}", assetTag, e.getMessage(), e);
            throw e;
        }
    }

    @Transactional(readOnly = true)
    public Page<Asset> listAssets(String query, Pageable pageable) {
        UUID orgId = getCurrentUserOrgId();
        log.debug("Listing assets for Org ID: {} with query: {}", orgId, query);
        if (query != null && !query.isBlank()) {
            return assetRepository.search(orgId, query.trim(), pageable);
        }
        return assetRepository.findByOrganization_Id(orgId, pageable);
    }

    @Transactional(readOnly = true)
    public Asset getAsset(UUID assetId) {
        log.debug("Fetching Asset ID: {}", assetId);
        Asset asset = assetRepository.findById(assetId)
                .orElseThrow(() -> new EntityNotFoundException("Asset not found"));
        ensureSameOrg(asset.getOrganization().getId());
        return asset;
    }

    public Asset updateAsset(UUID assetId, String assetType, String brand, String model,
            String serialNumber, LocalDate purchaseDate, LocalDate warrantyUntil,
            Boolean active, Integer totalQuantity) {
        log.info("Updating asset: {}", assetId);
        try {
            Asset asset = getAsset(assetId);

            if (assetType != null)
                asset.setAssetType(assetType);
            if (brand != null)
                asset.setBrand(brand);
            if (model != null)
                asset.setModel(model);
            if (serialNumber != null)
                asset.setSerialNumber(serialNumber);
            if (purchaseDate != null)
                asset.setPurchaseDate(purchaseDate);
            if (warrantyUntil != null)
                asset.setWarrantyUntil(warrantyUntil);
            if (active != null)
                asset.setActive(active);
            if (totalQuantity != null)
                asset.setTotalQuantity(totalQuantity);

            Asset saved = assetRepository.save(asset);
            log.info("Asset updated successfully: {}", assetId);
            return saved;
        } catch (Exception e) {
            log.error("Error updating asset {}: {}", assetId, e.getMessage(), e);
            throw e;
        }
    }

    public void deleteAsset(UUID assetId) {
        log.info("Deleting asset: {}", assetId);
        try {
            Asset asset = getAsset(assetId);
            assetRepository.delete(asset);
            log.info("Asset deleted successfully: {}", assetId);
        } catch (Exception e) {
            log.error("Error deleting asset {}: {}", assetId, e.getMessage(), e);
            throw e;
        }
    }

    // ============ ASSIGNMENT OPERATIONS ============

    public EmployeeAssetAssignment assignAsset(UUID userId, UUID assetId, LocalDate assignedOn,
            AssetCondition condition, String notes) {
        log.info("Assigning Asset ID: {} to User ID: {}", assetId, userId);
        try {
            Asset asset = getAsset(assetId);
            Employee employee = employeeRepository.findById(userId)
                    .orElseThrow(() -> new EntityNotFoundException("User not found: " + userId));

            // Ensure user is in same org as asset
            if (!employee.getOrganization().getId().equals(asset.getOrganization().getId())) {
                log.warn("Assignment failed: User {} and Asset {} not in same organization", userId, assetId);
                throw new AccessDeniedException("User and Asset must be in the same organization");
            }

            if (!asset.getActive()) {
                log.warn("Assignment failed: Asset {} is inactive", assetId);
                throw new IllegalArgumentException("Cannot assign an inactive asset");
            }

            long assignedCount = assignmentRepository.countByAsset_IdAndStatus(assetId, AssetAssignmentStatus.ASSIGNED);
            if (assignedCount >= asset.getTotalQuantity()) {
                log.warn("Assignment failed: Asset {} is out of stock", assetId);
                throw new IllegalArgumentException("Asset is out of stock");
            }

            EmployeeAssetAssignment assignment = EmployeeAssetAssignment.builder()
                    .employee(employee)
                    .asset(asset)
                    .assignedOn(assignedOn != null ? assignedOn : LocalDate.now())
                    .conditionOnAssign(condition != null ? condition : AssetCondition.GOOD)
                    .status(AssetAssignmentStatus.ASSIGNED)
                    .notes(notes)
                    .build();

            EmployeeAssetAssignment saved = assignmentRepository.save(assignment);
            log.info("Asset {} assigned successfully to user {}", assetId, userId);

            timelineService.createEvent(asset.getOrganization().getId(), "ASSET_ASSIGNMENT", saved.getId(), "ASSIGN",
                    "Asset Assigned", getCurrentUserId(), userId,
                    "Asset " + asset.getAssetTag() + " assigned to user", null);

            // Notify User
            notificationService.sendNotification(userId, "Asset Assigned",
                    "Asset " + asset.getBrand() + " " + asset.getModel() + " (" + asset.getAssetTag()
                            + ") has been assigned to you.",
                    "ASSET", saved.getId());

            return saved;
        } catch (Exception e) {
            log.error("Error assigning asset {} to user {}: {}", assetId, userId, e.getMessage(), e);
            throw e;
        }
    }

    public EmployeeAssetAssignment requestReturn(UUID assignmentId) {
        log.info("Requesting return for Assignment ID: {}", assignmentId);
        EmployeeAssetAssignment assignment = assignmentRepository.findById(assignmentId)
                .orElseThrow(() -> new EntityNotFoundException("Assignment not found"));

        UUID currentUserId = getCurrentUserId();
        if (!assignment.getEmployee().getId().equals(currentUserId)) {
            throw new AccessDeniedException("You can only request return for your own assigned assets");
        }

        if (assignment.getStatus() != AssetAssignmentStatus.ASSIGNED) {
            throw new IllegalArgumentException("Asset is not in ASSIGNED status");
        }

        assignment.setStatus(AssetAssignmentStatus.RETURN_REQUESTED);
        log.info("Return requested successfully for assignment {}", assignmentId);

        // Notify Org Admins
        notificationService.sendNotificationToOrgAdmins(assignment.getAsset().getOrganization().getId(), 
                "Asset Return Requested", 
                assignment.getEmployee().getFirstName() + " has requested to return asset " + assignment.getAsset().getAssetTag() + ".", 
                "ASSET", 
                assignment.getId());

        return assignmentRepository.save(assignment);
    }

    public EmployeeAssetAssignment confirmReturn(UUID assignmentId, LocalDate returnedOn,
            AssetCondition conditionOnReturn, String notes) {
        log.info("Confirming return for Assignment ID: {}", assignmentId);
        EmployeeAssetAssignment assignment = assignmentRepository.findById(assignmentId)
                .orElseThrow(() -> new EntityNotFoundException("Assignment not found"));

        ensureSameOrg(assignment.getAsset().getOrganization().getId());

        if (assignment.getStatus() != AssetAssignmentStatus.ASSIGNED &&
                assignment.getStatus() != AssetAssignmentStatus.RETURN_REQUESTED) {
            throw new IllegalArgumentException("Asset cannot be returned in current status");
        }

        LocalDate effectiveReturnDate = returnedOn != null ? returnedOn : LocalDate.now();
        if (effectiveReturnDate.isBefore(assignment.getAssignedOn())) {
            throw new IllegalArgumentException("Return date cannot be before assigned date");
        }

        assignment.setReturnedOn(effectiveReturnDate);
        assignment.setConditionOnReturn(conditionOnReturn != null ? conditionOnReturn : AssetCondition.GOOD);
        assignment.setStatus(AssetAssignmentStatus.RETURNED);
        if (notes != null)
            assignment.setNotes(notes);

        log.info("Asset return confirmed for assignment {}", assignmentId);

        timelineService.createEvent(assignment.getAsset().getOrganization().getId(), "ASSET_ASSIGNMENT", assignmentId,
                "RETURN", "Asset Returned", getCurrentUserId(), assignment.getEmployee().getId(),
                "Asset " + assignment.getAsset().getAssetTag() + " returned", null);

        // Notify User
        notificationService.sendNotification(assignment.getEmployee().getId(), "Asset Returned",
                "Return of asset " + assignment.getAsset().getBrand() + " " + assignment.getAsset().getModel()
                        + " (" + assignment.getAsset().getAssetTag() + ") confirmed.",
                "ASSET", assignment.getId());

        return assignmentRepository.save(assignment);
    }

    @Transactional(readOnly = true)
    public List<EmployeeAssetAssignment> getMyAssets() {
        UUID currentUserId = getCurrentUserId();
        log.debug("Fetching assets for Current User ID: {}", currentUserId);
        return assignmentRepository.findByEmployee_Id(currentUserId);
    }

    @Transactional(readOnly = true)
    public List<EmployeeAssetAssignment> getAssetHistory(UUID assetId) {
        log.debug("Fetching history for Asset ID: {}", assetId);
        getAsset(assetId); // Will check org access
        return assignmentRepository.findByAsset_Id(assetId);
    }

    // ============ HELPER METHODS ============

    private UUID getCurrentUserOrgId() {
        UserPrincipal principal = (UserPrincipal) SecurityContextHolder.getContext()
                .getAuthentication().getPrincipal();
        return principal.getOrgId();
    }

    private UUID getCurrentUserId() {
        UserPrincipal principal = (UserPrincipal) SecurityContextHolder.getContext()
                .getAuthentication().getPrincipal();
        return principal.getId();
    }

    private void ensureSameOrg(UUID targetOrgId) {
        UUID myOrgId = getCurrentUserOrgId();
        if (!myOrgId.equals(targetOrgId)) {
            throw new AccessDeniedException("Access denied: different organization");
        }
    }

    // Helper to get available quantity
    public int getAvailableQuantity(UUID assetId) {
        log.trace("Calculating available quantity for Asset ID: {}", assetId);
        Asset asset = assetRepository.findById(assetId)
                .orElseThrow(() -> new EntityNotFoundException("Asset not found"));
        long assigned = assignmentRepository.countByAsset_IdAndStatus(assetId, AssetAssignmentStatus.ASSIGNED);
        return asset.getTotalQuantity() - (int) assigned;
    }
}
